src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -351,3 +367,419 @@ deriving stock instance Read SizeRatio
 deriving stock instance Read MergePolicy
 deriving stock instance Read BloomFilterAlloc
 deriving stock instance Read FencePointerIndex
+
+{-------------------------------------------------------------------------------
+  Encoding and decoding
+-------------------------------------------------------------------------------}
+
+class Encode a where
+  encode :: a -> Encoding
+
+class Decode a where
+  decode :: Decoder s a
+
+class DecodeVersioned a where
+  decodeVersioned :: SnapshotVersion -> Decoder s a
+
+encodeSnapshotMetaData :: SnapshotMetaData -> ByteString
+encodeSnapshotMetaData = toLazyByteString . encode . Versioned
+
+decodeSnapshotMetaData :: ByteString -> Either DeserialiseFailure SnapshotMetaData
+decodeSnapshotMetaData bs = second (getVersioned . snd) (deserialiseFromBytes decode bs)
+
+newtype Versioned a = Versioned { getVersioned :: a }
+  deriving stock (Show, Eq)
+
+instance Encode a => Encode (Versioned a) where
+  encode (Versioned x) =
+       encodeListLen 2
+    <> encode currentSnapshotVersion
+    <> encode x
+
+instance DecodeVersioned a => Decode (Versioned a) where
```

Comment:
And this is where we transition from `Decode` to `DecodeVersioned`, with the `Versioned` wrapper that stores the version. And everything below gets decoded with access to that version.